azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -223,73 +223,141 @@ def _handle_pipeline_event(self, event):
                     self.name, op.name
                 )
             )
             self.queue.put_nowait(op)
 
-        # If we get a request to connect, we either complete immediately (if we're already
-        # connected) or we do the connect operation, which is pulled out into a helper
-        # function because starting the connection also means blocking this stage until
-        # the connect is complete.
-        elif isinstance(op, pipeline_ops_base.ConnectOperation):
-            if self.connected:
+        elif isinstance(op, pipeline_ops_base.ConnectOperation) and self.connected:
+            logger.info(
+                "{}({}): Transport is already connected.  Completing early".format(
+                    self.name, op.name
+                )
+            )
+            operation_flow.complete_op(stage=self, op=op)
+
+        elif isinstance(op, pipeline_ops_base.DisconnectOperation) and not self.connected:
+            logger.info(
+                "{}({}): Transport is already disconnected.  Completing early".format(
+                    self.name, op.name
+                )
+            )
+            operation_flow.complete_op(stage=self, op=op)
+
+        elif (
+            isinstance(op, pipeline_ops_base.DisconnectOperation)
+            or isinstance(op, pipeline_ops_base.ConnectOperation)
+            or isinstance(op, pipeline_ops_base.ReconnectOperation)
+        ):
+            self._block(op)
+            old_callback = op.callback
+
+            @pipeline_thread.runs_on_pipeline_thread
+            def on_operation_complete(op):
                 logger.info(
-                    "{}({}): protocol client is already conencted.  completing early.".format(
-                        self.name, op.name
+                    "{}({}): complete.  Unblocking queue with error: {}".format(
+                        self.name, op.name, op.error
                     )
                 )
-                operation_flow.complete_op(self, op)
-            else:
-                self._do_connect(op)
-
-        # If we get a request to disconnect, we either complete the request immediately
-        # (if we're already disconencted) or we pass the disconnect request down.
-        elif isinstance(op, pipeline_ops_base.DisconnectOperation):
-            if not self.connected:
+                op.callback = old_callback
+                self._unblock(op, op.error)
                 logger.info(
-                    "{}({}): procotol client is already disconencted.  completing early.".format(
+                    "{}({}): unblock is comlete.  completing op that caused unblock".format(
                         self.name, op.name
                     )
                 )
-                operation_flow.complete_op(self, op)
-            else:
-                operation_flow.pass_op_to_next_stage(self, op)
+                operation_flow.complete_op(stage=self, op=op)
```

Comment:
>operation_flow.complete_op(stage=self, op=op) [](start = 16, length = 45)

Doesn't this mean that the operation gets "completed" twice?

It gets completed below, thus triggering this callback, and then completes again here in the callback, in order to complete the ORIGINAL callback. This seems odd, will generate potentially confusing logs, and could result in issues if "complete_op" ever does more than just triggering a callback.

Reply:
That's exactly what it means, and you're right.  The other option is to copy every single operation that goes down the pipeline.  I'll have to think about this more, I didn't even consider this as a point of confusion because it's a fairly common practice in driver development (though with different terms).

---
In reply to: [311710166](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311710166) [](ancestors = 311710166)